Clean up run_cmake_test

JessicaMeixner-NOAA opened the pull request:
# Pull Request Summary
Provides clean up and performance improvements for regression tests with cmake

## Description
This PR: 
* cleans up the options in run_cmake_test to remove unused options
* ensures that we are not building a MPI/SHRD version of the code unless we actually are using a mpi or omp option (essentially were double building before)  
* Optionally uses a "path_build_root" in run_cmake test which allows for multiple build directories if running multiple regression tests and allows for re-use of build directories. 
* Fixes the timing -T option in the run_cmake_test to properly count compile time which was broken

Please also include the following information: 
* Add any suggestions for a reviewer @ukmo-ccbunney @mickaelaccensi @thesser1 (Please tests your various matrix_cmake_* submit scripts) 
* Mention any labels that should be added:   _enhancement_
* Are answer changes expected from this PR? no 

### Issue(s) addressed
* Fixes #49
Might fix #687 needs confirmation though 

### Commit Message

Enhancement and clean-up of run_cmake_test


### Check list  

<!-- After creating the PR you can check each of the items below that have been completed -->

- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [ ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested? matrix_cmake_test 
* Are the changes covered by regression tests? Yes
* Have the matrix regression tests been run (if yes, please note HPC and compiler)? Yes, ncep intel
* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).  -- no expected changes (the list of differences have recently changed though, this gives me the same diffs as comparing two sets or runs from the develop branch): 
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (17 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
ww3_ta1/./work_UPD0F_U                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.17/./work_c                     (1 files differ)
ww3_tp2.17/./work_b                     (1 files differ)
ww3_tp2.17/./work_a                     (1 files differ)
ww3_tp2.6/./work_ST0                     (1 files differ)
ww3_tp2.6/./work_ST4                     (1 files differ)
ww3_tp2.6/./work_pdlib                     (1 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```

* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/8620302/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/8620303/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/8620304/matrixDiff.txt)

JessicaMeixner-NOAA committed:
clean up of run_cmake_test

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into cleanup

JessicaMeixner-NOAA committed:
updates to run_cmake_test to run faster
and for matrix divider to have two integer digits

JessicaMeixner-NOAA committed:
update run_cmake_test

JessicaMeixner-NOAA committed:
clean up build directories

JessicaMeixner-NOAA committed:
fix tp2.14 prep_env

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/cleanup' into feature/builddirforrt

JessicaMeixner-NOAA committed:
delete all build directories

JessicaMeixner-NOAA committed:
remove cmplr from options in ukmo

JessicaMeixner-NOAA commented:
@ukmo-ccbunney I noticed in matrix_cmake_ukmo_cray you are still using w3_setenv, in matrix_cmake_ncep we require an input of the model path.  w3_setenv will likely go away with the clean-up of the gnu build, do you want me to add the model dir input to your script here or leave it as is?

aliabdolali commented on regtests/bin/matrix_cmake_ncep:
```diff
@@ -43,11 +43,8 @@ fi
 
-# Compiler option. Choose appropriate compiler and set cmplOption to
-# y if using for the first time or using a different compiler
+#download 
```
do we need this comment?

JessicaMeixner-NOAA replied:
No we do not need this anymore.  Good catch!

JessicaMeixner-NOAA replied:
This has been updated and i cleaned up some other formatting while i was double checking things.

JessicaMeixner-NOAA committed:
comment cleanup

mickaelaccensi reviewed (approved):
regtests passed

ukmo-ccbunney commented:
> @ukmo-ccbunney I noticed in matrix_cmake_ukmo_cray you are still using w3_setenv, in matrix_cmake_ncep we require an input of the model path. w3_setenv will likely go away with the clean-up of the gnu build, do you want me to add the model dir input to your script here or leave it as is?

Yes please. I have made the change locally and am testing your new changes.

JessicaMeixner-NOAA committed:
update matrix_cmake_ukmo_cray

JessicaMeixner-NOAA commented:
@ukmo-ccbunney I added the update to grab the model path from the arguments.  I also added the matrix divider at the end since I noticed that was missing... not sure if you want that added though?

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/builddirforrt

ukmo-ccbunney commented:
> @ukmo-ccbunney I added the update to grab the model path from the arguments. I also added the matrix divider at the end since I noticed that was missing... not sure if you want that added though?

Thanks.
I usually run the matrix_divider manually, but I am happy for it to be included in the matrix generation script.
I should have results of regtests very soon.

aliabdolali commented:
@JessicaMeixner-NOAA thanks for doing all of this. 
@ukmo-ccbunney I am going to test, and do my review, once I have your approval, we can proceed with this PR.

ukmo-ccbunney commented:
@JessicaMeixner-NOAA I notice that in matrix_cmake_ncep you don't explicitly set any compiler anymore. How can you be sure that CMake is picking up the compiler you want (intel, I assume)?

Are you relying on what is in the FC environment variable?

ukmo-ccbunney commented:
Apologies for the delay from my end - I am getting some differences in `*_trck.nc` and `*_spec.nc` files across multiple regtetst that I am trying to get to the bottom of (tiny differences in the `freq` arrays in the netCDF files).

ukmo-ccbunney reviewed (approved):
OK - after checking two fresh copies of develop and this branch, the `mww3_test_02` regtests are now b4b. I have no idea what was happening previously to cause the differences...

Apologies for the delay - all good now!

aliabdolali commented:
All tests passed on NOAA hpc with pre-known non-identical cases

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (17 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (17 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
ww3_ta1/./work_UPD0F_U                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (6 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)

```
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/8671509/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/8671510/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/8671511/matrixDiff.txt)

JessicaMeixner-NOAA commented:
> @JessicaMeixner-NOAA I notice that in matrix_cmake_ncep you don't explicitly set any compiler anymore. How can you be sure that CMake is picking up the compiler you want (intel, I assume)?
> 
> Are you relying on what is in the FC environment variable?

@ukmo-ccbunney we're relying on the modules to pick the compiler.. I believe this then uses the environment variables but then how/where that happens that's a great question...
